emitter.go

```diff
@@ -70,17 +67,22 @@ func (e *Emitter) EmitWithHeaders(key string, msg interface{}, headers map[strin
 			return nil, fmt.Errorf("Error encoding value for key %s in topic %s: %v", key, e.topic, err)
 		}
 	}
-	e.wg.Add(1)
-	if headers == nil {
-		return e.producer.Emit(e.topic, key, data).Then(func(err error) {
-			e.wg.Done()
-		}), nil
-	} else {
-		return e.producer.EmitWithHeaders(e.topic, key, data, headers).Then(func(err error) {
-			e.wg.Done()
-		}), nil
+
+	// protect e.done channel and e.wg WaitGroup together to reject all new emits after calling e.Finish
+	e.mu.Lock()
```

Comment:
I think we should use an RWMutex and do e.mu.RLock() here, otherwise we introduce a synchronization point and goroutines will not be able to Emit messages concurrently

Reply:
I think tried `&atomic.AddInt64(&count, 1)` at first, which still gave me data races and I though a mutext would be too much. But maybe i missed something